pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -783,6 +789,117 @@ func setSoMarks(fd int, ipFamily ipfamily.IPFamily, secId identity.NumericIdenti
 	return nil
 }
 
+var errNoReader = errors.New("Reader stopped")
+
+type Response struct {
+	*dns.Msg
+	err error
+}
+
+// A Client keeps state for concurrent transactions on the same upstream connection.
+type Client struct {
+	dns.Client
+	refcount int // protected by proxy's lock
+
+	// Client's connection shared among all requests from the same source address/port. The
+	// locks below is used to serialize reads and writes on this connection, but reads and
+	// writes can happen at the same time.
+	conn *dns.Conn
+
+	// this mutex protects writes on 'conn' and all access to 'reqs'
+	lock.RWMutex
+	reqs map[uint16]chan Response // outstanding requests
+
+	// 'reader' mutex is used to serialize reads on 'conn'. It is always taken and released
+	// while holding the lock but the lock can be released and re-acquired while holding
+	// 'reader' mutex.
+	reader lock.Mutex
+}
+
+// ExchangeAsync writes the request to the Client's connection and co-operatively
+// reads responses from the connection and distributes them to the requestors.
+// At most one caller is reading from Client's connection at any time.
+func (c *Client) ExchangeAsync(m *dns.Msg) (r *dns.Msg, rtt time.Duration, err error) {
+	// Lock allows only one request to be written at a time, but that can happen
+	// concurrently with reading.
+	c.Lock()
+	defer c.Unlock()
+	if _, exists := c.reqs[m.Id]; exists {
+		return nil, 0, fmt.Errorf("Duplicate request: %d", m.Id)
+	}
+
+	// Send while holding the client lock, as dns.Client is not made to be usable from
+	// concurrent goroutines.
+	start := time.Now()
+	err = c.SendContext(context.Background(), m, c.conn, start)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Create channel for the response with buffer of one, so that write to it
+	// does not block if we happen to do it ourselves.
+	ch := make(chan Response, 1)
+	c.reqs[m.Id] = ch
+
+	// Wait for the response
+	var resp Response
+	for {
+		// Try taking the reader lock
+		if c.reader.TryLock() {
```

Comment:
As the docs say: `Note that while correct uses of TryLock do exist, they are rare, and use of TryLock is often a sign of a deeper problem in a particular use of mutexes.`

I'm not yet convinced that this is such a case. The code is somewhat complicated, and reasoning about it in terms of deadlocks or race conditions thus difficult.

Could you explain the reasoning behind taking this approach a bit more? On what grounds do you think caching the socket is worth the additional complexity here? Or is the intention to make the proxy transparent the intent of the PR, and this complexity is just the fall-out?

Reply:
Yes, the intention is to make the DNS proxy transparent to make it play nice with IPsec. As a consequence we must support multiple requests on the same connection (5-tuple), as the DNS clients are doing it (e.g. curl without the `-4` flag does `A` and `AAAA` queries from the same socket back-to-back without waiting for a response).

I first tried an approach where each DNS client is served by a goroutine, but also adds more complexity. While being easier to reason about locking behavior, it adds the overhead of an additional goroutine to each DNS request (minus 5-tuple reuse).

Looking at the `cilium/dns` code I think it is not safe to allow more than one goroutine read from a client's connection as there is state that is not protected against concurrent access. I tried to keep `cilium/dns` changes to a minimum to not expand the scope of this PR too much, so I went with the current approach, where the control flow is almost the same as before, with the addition of passing responses to waiting requestors and a remaining waiter taking over the reading once the current reader gets its own response.

(in fact at first I was just running multiple clients on the same 5-tuple which was possible due to reuseport option, but that resulted in a wrong client reading a response and trashing it for a non-matching ID. This would likely have failed for TCP as well.)

All this being said I'm all open for ideas to make this better :-)

Reply:
@bimmlerd It should be noted that each `curl` invocation in a test with toFQDNs policy stresses this concurrency pattern due to the back-to-back `A` and `AAAA` requests I mentioned above.

Reply:
Also, the read/write deadlines set on the connection will make the blocking read error out on timeout, so the read loop will terminate on that.

Reply:
Also, we do not know about the need for the shared client until the 2nd request comes in while the 1st response is still pending. At that point it is too late to schedule a shared reader (via a goroutine) as the 1st requestor is already blocking on the read. Hence we would need to schedule a goroutine already when the 1st request is being served, which adds per-request overhead.

We could do a more involving redesign for UDP where we would only use one raw UDP socket with per-packet control block for controlling the local address, like we already do for getting the UDP requests in in the first place. In that case we could still need a goroutine per TCP connection, unless we do this kind of "co-operative" reading from the shared client connection.

Reply:
For reasoning about locking:
- `c.reader.TryLock` can no block
- there is only one other lock in this pattern, therefore `c.Lock()` can not deadlock, unless there is a path where a goroutine does not eventually call `c.Unlock()`. We do `defer c.Unlock()` so this is highly unlikely
- I had this pattern deadlock on development, though, on write to the reader's own channel. I made the channel buffered for that reason. If there is a path where there are more than one write to any of the `reqs` channels, there could be a deadlock. This is why we delete the channel from the map before writing on the channel, which makes multiple writes to any one channel also impossible

Reply:
Did you think about which locks are held outside of the DNS proxy when control flow enters these critical sections? Specifically, are we allowed to hold the DNSProxy lock here, or does that create a potential wait cycle involving Endpoints & ipcache?